libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -0,0 +1,235 @@
+
+        RTTR_FOREACH_PT(MapPoint, map.size)
+        {
+            if(waterDistance[pt] > 0 && mountainDistance[pt] == 0)
+            {
+                waterDistance[pt] = 1;
+            }
+        }
+
+        auto range = waterDistance.GetRange();
+        auto probRange = ValueRange<unsigned>(15, 40);
+        auto probDiff = probRange.GetDifference();
+
+        auto mountainDepth =
+          Distances(map.size, [&map](const MapPoint& pt) { return !map.textures.All(pt, IsMountainOrSnowOrLava); });
+
+        auto maximumMountainDepth = mountainDepth.GetMaximum();
+        auto mountainRange = mountainDepth.GetRange();
+
+        RTTR_FOREACH_PT(MapPoint, map.size)
+        {
+            const unsigned waterDist = waterDistance[pt];
+            const unsigned mountainDist = mountainDistance[pt];
+
+            if(waterDist > 0 && distanceToForbiddenArea[pt] > 0)
+            {
+                if(mountainDist > 0)
+                {
+                    auto prob = MapValueToIndex(waterDist, range, probDiff);
+                    probabilities[pt] = prob + probRange.minimum;
+                } else
+                {
+                    auto diff = maximumMountainDepth - mountainDepth[pt];
+                    auto prob = MapValueToIndex(diff, mountainRange, probDiff);
+                    probabilities[pt] = prob + probRange.minimum;
+                }
+            } else
+            {
+                probabilities[pt] = 0;
+            }
+        }
+
+        auto trees = CreateTrees(map.textures);
+        auto treeForPoint = [&mountainDistance, &waterDistance, &range, &trees](const MapPoint& pt) {
+            if(mountainDistance[pt] == 0)
+            {
+                return trees.back();
+            }
+            auto index = MapValueToIndex(waterDistance[pt], range, trees.size());
+            return trees[index];
+        };
+
+        RTTR_FOREACH_PT(MapPoint, map.size)
+        {
+            auto treeProb = probabilities[pt];
+
+            if(treeProb > 0 && rnd.ByChance(treeProb))
+            {
+                auto tree = treeForPoint(pt);
+
+                map.objectInfos[pt] = tree.type;
+                map.objectTypes[pt] = tree.index + rnd.RandomInt(0, 7);
+            }
+
+            auto graniteProb = probabilities[pt] / 4;
+
+            if(graniteProb > 0 && rnd.ByChance(graniteProb))
+            {
+                map.objectInfos[pt] = rnd.ByChance(50) ? 0xCC : 0xCD;
+                map.objectTypes[pt] = rnd.RandomInt(1, 6);
+            }
+        }
+    }
+
+    void AddResources(Map& map, RandomUtility& rnd, const MapSettings& settings)
+    {
+        const auto& textures = map.textures;
+        auto& resources = map.resources;
+
+        RTTR_FOREACH_PT(MapPoint, map.size)
+        {
+            if(textures.All(pt, IsMinableMountain))
+            {
+                int randomNumber = rnd.RandomInt(1, 100);
+                int ratio = settings.ratioGold;
```

Comment:
Wouldn't this lead to clusters of gold in 1 corner and stone in the other?

Reply:
fixed